noir-circuits/blob/src/main.nr

```diff
@@ -161,11 +48,11 @@ fn unsafe_blob_to_fields(blob: [F; FIELDS_PER_BLOB]) -> [Field; FIELDS_PER_BLOB]
 //
 // For efficiency, the top_bit is kept as a Field throughout.
 fn unsafe_extract_top_bit(x: F) -> (Field, F) {
-    let top_limb: Field = x.limbs[3];
-    // The top_limb is at most 2 bytes (16 bits).
-    let top_bit: Field = (top_limb as u16 / 0x8000) as Field;
-    let top_limb_with_top_bit_removed = top_limb - top_bit * 0x8000;
-    let x_with_top_bit_removed: F = BigNum { limbs: [x.limbs[0], x.limbs[1], top_limb_with_top_bit_removed] };
+    let (top_bit, x_with_top_bit_removed) = __unsafe_extract_top_bit(x);
+    assert_eq(top_bit * 0x8000 + x_with_top_bit_removed.limbs[2], x.limbs[2]);
+    assert_eq(x_with_top_bit_removed.limbs[1], x.limbs[1]); // BUG: this should be free; it's currently contributing 4k constraints
+    assert_eq(x_with_top_bit_removed.limbs[0], x.limbs[0]); // BUG: this should be free; it's currently contributing 4k constraints
```

Comment:
It's not clear to me why this should be free. You're returning a value from an unconstrained function (which can be anything) and asserting that it is equal to a value in the circuit.

We can't trust that the prover will return the limbs from `x` as specified in the unconstrained function so we cannot optimise out these constraints.

Reply:
In this situation it would be best to not return an `F` from `__unsafe_extract_top_bit` but just the top bit and top limb. You can then construct the new `F` in a constrained function more cheaply than doing it an unconstrained function and making assertions about it.

Reply:
```rust
unconstrained fn __unsafe_extract_top_bit(x: F) -> (Field, Field) {
    let top_limb: Field = x.limbs[2];
    // The top_limb is at most 2 bytes (16 bits).
    // 0x8000 = 2^15 = 32768
    let top_bit: Field = (top_limb as u16 / 0x8000) as Field;
    let top_limb_with_top_bit_removed = top_limb - top_bit * 0x8000;
    (top_bit, top_limb_with_top_bit_removed)
}

// DANGER: it's named as "unsafe" because the caller MUST already have checked that 
// each blob Field is formatted as (u1, Field). I.e. the "rhs" 254-bits should already
// fit within a Field. If the "rhs" 254 bits is larger than the field modulus,
// there will be an uncaught overflow of the 254-bits in the Field, resulting in 
// an unintended tiny value.
//
// For efficiency, the top_bit is kept as a Field throughout.
fn unsafe_extract_top_bit(x: F) -> (Field, F) {
    let (top_bit, top_limb_with_top_bit_removed) = __unsafe_extract_top_bit(x);
    assert_eq(top_bit * 0x8000 + x_with_top_bit_removed.limbs[2], x.limbs[2]);

    (top_bit, BigNum { limbs: [x.limbs[0], x.limbs[1], top_limb_with_top_bit_removed] })
}

```

Reply:
I thought asserting two witnesses to be equal can be done with a copy-constraint, which just leverages the permutation polynomial stuff of plonk, instead of creating a gate?
I _think_ that's what bberg does. It does a weird thing where it actually just sets one of the arguments of `assert_equal` to have the same witness index as the other, iirc.

Maybe I'm misremembering because I've been out of the weeds for 2 years.

Thanks for taking the time to suggest an improvement here :)

Reply:
This is something that we can maybe do in the DSL folder but in Noir this is just an AssertZero opcode like any other.

Reply:
Generally though if you want continuity it's best to do that entirely in constrained-land.